File: js/1차 다트게임(2018 kakao).js
function solution(dartResult) {
    const matches = dartResult.match(/(\d+)([SDT])([*#]?)/g);
    var arr = [];
    
    for (const m of matches) {
        // 정규식은 3개의 캡처 그룹(()로 감싸진 부분)을 가지고 있으므로, match는 다음과 같은 배열을 반환
        // ["전체 매칭된 문자열", "첫 번째 캡처 그룹", "두 번째 캡처 그룹", "세 번째 캡처 그룹"]
        // 전체 매칭된 문자열 (match[0])이 필요하지 않다면 다음과 같이 무시할 수 있음
        // [_, 숫자, 보너스, 옵션]
        var [_, point, bonus, option] = m.match(/(\d+)([SDT])([*#]?)/);
        point = parseInt(point);
        if (bonus === 'D') {
            point **= 2;
        } else if (bonus === 'T') {
            point **= 3;
        }
        
        if (option === '*') {
            if (arr) {
                arr[arr.length-1] *= 2;
            }
            point *= 2;
        } else if (option === '#') {
            point *= -1;
        }
        arr.push(point);
    }
    return arr.reduce((acc,cur) => acc+cur, 0);
}

File: python/1차 다트게임(2018 kakao).py
import re

def solution(dartResult):
    arr = [0]
    divide_by_num_arr = re.findall(r'\d+[A-Z*#]+', dartResult);
    for elem in divide_by_num_arr:
        if len(elem) == 3:
            if elem[2] == '*':
                arr[-1] *= 2
                if elem[1] == 'S':
                    arr.append(int(elem[0]) * 2)
                elif elem[1] == 'D':
                    arr.append((int(elem[0]) **2) * 2)
                else:
                    arr.append((int(elem[0]) ** 3) * 2)
                    continue
            elif elem[2] == '#':
                if elem[1] == 'S':
                    arr.append(int(elem[0]) - int(elem[0]) * 2)
                elif elem[1] == 'D':
                    arr.append(int(elem[0]) ** 2 - (int(elem[0]) ** 2) * 2)
                else:
                    arr.append(int(elem[0]) ** 3 - (int(elem[0]) ** 3) * 2)
            else:
                if elem[2] == 'S':
                    arr.append(int(elem[0:2]))
                elif elem[2] == 'D':
                    arr.append(int(elem[0:2]) ** 2)
                else:
                    arr.append(int(elem[0:2]) ** 3)
        else:
            if elem[1] == 'S':
                arr.append(int(elem[0]))
            elif elem[1] == 'D':
                arr.append(int(elem[0]) ** 2)
            else:
                arr.append(int(elem[0]) ** 3)
    return sum(arr)

# S = 1제곱
# D = 2제곱
# T = 3제곱

# 옵션
# * = 해당 점수와 바로 전 점수 2배
# # = 해당 점수 마이너스
